libyul/optimiser/StackLimitEvader.cpp

```diff
@@ -0,0 +1,141 @@
+
+		// Assign to zero early to guard against recursive calls.
+		nextAvailableSlot[_function] = 0;
+
+		uint64_t nextSlot = 0;
+		if (callGraph.count(_function))
+			for (YulString child: callGraph.at(_function))
+				nextSlot = std::max(run(child), nextSlot);
+
+		if (unreachableVariables.count(_function))
+		{
+			yulAssert(!slotAllocations.count(_function), "");
+			auto& assignedSlots = slotAllocations[_function];
+			for (YulString variable: unreachableVariables.at(_function))
+				if (variable.empty())
+				{
+					// TODO: Too many function arguments or return parameters.
+				}
+				else
+					assignedSlots[variable] = nextSlot++;
+		}
+
+		return (nextAvailableSlot[_function] = nextSlot);
+	}
+
+	map<YulString, map<YulString, uint64_t>> slotAllocations{};
+	map<YulString, uint64_t> nextAvailableSlot{};
+};
+u256 getLiteralArgumentValue(FunctionCall const& _call)
+{
+	yulAssert(_call.arguments.size() == 1, "");
+	Literal const* literal = std::get_if<Literal>(&_call.arguments.front());
+	yulAssert(literal && literal->kind == LiteralKind::Number, "");
+	return valueOfLiteral(*literal);
+}
+}
+
+void StackLimitEvader::run(
+	OptimiserStepContext& _context,
+	Object& _object,
+	map<YulString, set<YulString>> const& _unreachableVariables
+)
+{
+	yulAssert(_object.code, "");
+	auto const* evmDialect = dynamic_cast<EVMDialect const*>(&_context.dialect);
+	yulAssert(
+		evmDialect && evmDialect->providesObjectAccess(),
+		"StackToMemoryMover can only be run on objects using the EVMDialect with object access."
+	);
+
+	vector<FunctionCall*> memoryGuardCalls = FunctionCallFinder::run(
+		*_object.code,
+		"memoryguard"_yulstring
+	);
+	// Do not optimise, if no ``memoryguard`` call is found.
+	if (memoryGuardCalls.empty())
+		return;
+
+	// Make sure all calls to ``memoryguard`` we found have the same value as argument (otherwise, abort).
+	u256 reservedMemory = getLiteralArgumentValue(*memoryGuardCalls.front());
+	for (FunctionCall const* getFreeMemoryStartCall: memoryGuardCalls)
+			if (reservedMemory != getLiteralArgumentValue(*getFreeMemoryStartCall))
+				return;
+
+	CallGraph callGraph = CallGraphGenerator::callGraph(*_object.code);
+
+	// We cannot move variables in recursive functions to fixed memory offsets.
+	for (YulString function: callGraph.recursiveFunctions())
+		if (_unreachableVariables.count(function))
+			return;
+
+	MemoryOffsetAllocator memoryOffsetAllocator{_unreachableVariables, callGraph.functionCalls};
+	uint64_t requiredSlots = memoryOffsetAllocator.run();
+
+	StackToMemoryMover{_context, reservedMemory, memoryOffsetAllocator.slotAllocations}(*_object.code);
+	reservedMemory += 32 * requiredSlots;
+	YulString reservedMemoryString{util::toCompactHexWithPrefix(reservedMemory)};
+	for (FunctionCall* memoryGuardCall: memoryGuardCalls)
+	{
+		Literal* literal = std::get_if<Literal>(&memoryGuardCall->arguments.front());
+		yulAssert(literal && literal->kind == LiteralKind::Number, "");
+		literal->value = reservedMemoryString;
```

Comment:
I get what this is doing, but can you briefly explain the concept behind it? (Just as a GH comment)

Reply:
Yeah, the ``memoryguard`` thing takes some getting used to... this restores the semantics of it :-). So the idea is this:

- Before the stack limit evader you have e.g. ``let ptr := memoryguard(0x80)``, i.e. the code promises to use only memory in ``[0, 0x80)`` and ``[ptr, infinity)``.
- The stack limit evader now may decide that it needs ``0x0100`` bytes of memory. So it recognizes that so far only ``[0, 0x80)`` was reserved and starts assigning slots between ``0x80`` and ``0x0180``.
- Now it actually moves variables to ``[0x80, 0x0180)`` and does so by using plain ``mload``s and ``mstore``s.
- So you end up with code that now reserves ``[0, 0x0180)`` and not only ``[0, 0x80)``.
- So to preserve the semantics of ``memoryguard`` you have to change the argument from ``0x80`` to ``0x0180``.

This way after the optimiser step is run, you are in the same situation as before, just that the code has reserved a bit more of memory. This way, if for whatever reason you run this step again or any other step that might need memory in the future, that next step will correctly assume that ``[0, 0x0180)`` is now reserved.

Reply:
Ok!